### ISSUE TITLE
I can't generate meaningful results on datasets like: LSUN and ImageNet

### ISSUE THREAD

**Issue by Josh00-Lu** [AlexGraikos/diffusion_priors#1]

**Comment by AlexGraikos** [comment]
Hi, can you elaborate more on what you have tried and what did not work? 

- Did you train the model on LSUN and ImageNet yourself? 
- Can you generate unconditional samples with the original DDPM algorithm? 
- What do you mean by meaningful? I expect that generating samples with our algorithm on datasets like ImageNet may give you modes that are not realistic-looking.

**Comment by Josh00-Lu** [comment]
Thanks for your reply!

1) I use the pretrained DDPM backbone provided by this repo: https://github.com/openai/guided-diffusion. 
2) I can generate unconditional samples with the original DDPM. 
3) The script I use is based on the .ipynb file provided in folder "FFHQ".
```
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
import math

from unet import UNetModel
from diffusion import GaussianDiffusion

import tqdm
import matplotlib.pyplot as plt

# Load DDPM model
device = torch.device('cuda:0')

diff_net = UNetModel(image_size=256, in_channels=3, out_channels=6, 
                     model_channels=256, num_res_blocks=2, channel_mult=(1, 1, 2, 2, 4, 4),
                     attention_resolutions=[32,16,8], num_head_channels=64, dropout=0.1, resblock_updown=True, use_scale_shift_norm=True).to(device)
diff_net.load_state_dict(torch.load('models/ffhq.pt'))
print('Loaded Diffusion Model')

X0 = torch.randn(1, 3, 256, 256).to(device) 
X0.requires_grad_(True)
# Inference procedure steps
steps = 250   

opt = torch.optim.Adamax([X0], lr=1)

diffusion = GaussianDiffusion(T=1000, schedule='linear')
diff_net.eval()

bar = tqdm.tqdm(range(steps))
for i, _ in enumerate(bar):
    # Select t      
    t = ((steps-i)/1.5 + (steps-i)/3*math.cos(i/10))/steps*800 + 200 # Linearly decreasing + cosine
    t = np.array([t + np.random.randint(-50, 51) for _ in range(1)]).astype(int) # Add noise to t
    t = np.clip(t, 1, diffusion.T)
       
    # Denoise
    sample_img = X0
    xt, epsilon = diffusion.sample(sample_img, t)       
    t = torch.from_numpy(t).float().view(1)    
    pred = diff_net(xt.float(), t.to(device))   
    epsilon_pred = pred[:,:3,:,:] # Use predicted noise only
    
    loss = F.mse_loss(epsilon_pred, epsilon)
    bar.set_description("Loss: {:.4f}".format(loss))

    opt.zero_grad()        
    loss.backward()
                  
    opt.step()

    if (i+1) % 25 == 0 or i == 0:
        with torch.no_grad():
            fig, ax = plt.subplots(1, 1, figsize=(10,5))
            ax.imshow(0.5*(X0+1)[0].cpu().numpy().transpose([1,2,0]))
            ax.set_title('Inferred Image')
            plt.show()
```

Here's my sampled results using play-and-plug on ImageNet:
```
device = torch.device('cuda:0')

diff_net = UNetModel(image_size=256, in_channels=3, out_channels=6,
                     model_channels=256, num_res_blocks=2, channel_mult=(1, 1, 2, 2, 4, 4),
                     attention_resolutions=[32,16,8], num_head_channels=64, dropout=0.0, resblock_updown=True, use_scale_shift_norm=True).to(device)
diff_net.load_state_dict(torch.load('models/256x256_diffusion_uncond.pt'))
print('Loaded Diffusion Model')
```

![image](https://user-images.githubusercontent.com/74221638/196340567-5b38928f-a0c2-4008-af67-6b657d3ecd3a.png)
![image](https://user-images.githubusercontent.com/74221638/196340635-c571f301-539d-496e-85ea-82f77db2be8c.png)

However, with the same code, results are pretty good on FFHQ:
![image](https://user-images.githubusercontent.com/74221638/196341100-57a8993f-61c1-4127-94c2-d18cda8e175d.png)
![image](https://user-images.githubusercontent.com/74221638/196342906-85ef860e-2ba8-4433-9586-a81c2c07697d.png)


Have you tried your methods on unaligned dataset like ImageNet? or are there any restrictions on dataset choosing?

Hope for your further reply!

**Comment by AlexGraikos** [comment]
We did not try it on unaligned datasets like ImageNet and my guess is that the more complex the dataset, the harder it is to find modes that are likely to look like realistic samples. 

However, if I try to generate with a conditioning signal from an auxiliary classifier, I seem to be able to start moving towards some meaningful images. In this case, I took their unconditional ImageNet model and classifier (using the latter only for t=1) and by running the first few steps of optimization I can see the sample shifting towards some image from the subset I selected.

Volcano:
![image](https://user-images.githubusercontent.com/11280649/196570081-2a220e44-d26c-4dc1-b693-153bf5f3670c.png)

Acoustic Guitar:
![image](https://user-images.githubusercontent.com/11280649/196572053-3e19adbf-f97b-4e11-97ac-0884c8ce12fe.png)

How to continue this process to generate the final samples is open to discussion.
